### PR TITLE
Have router make two separate attempts to reconnect to current partition.

### PR DESCRIPTION
**Pull request by jwhui**

### CODE REVIEW

File: include/openthread-types.h [hunk 1]
```diff
@@ -691,8 +691,9 @@ typedef struct otExternalRouteConfig
 typedef enum otMleAttachFilter
 {
     kMleAttachAnyPartition    = 0,  ///< Attach to any Thread partition.
-    kMleAttachSamePartition   = 1,  ///< Attach to the same Thread partition.
-    kMleAttachBetterPartition = 2,  ///< Attach to a better (i.e. higher weight/partition id) Thread partition.
+    kMleAttachSamePartition1  = 1,  ///< Attach to the same Thread partition (attempt 1).
+    kMleAttachSamePartition2  = 2,  ///< Attach to the same Thread partition (attempt 2).
+    kMleAttachBetterPartition = 3,  ///< Attach to a better (i.e. higher weight/partition id) Thread partition.
```
Comment (librasungirl) [inline include/openthread-types.h]: should we also define `kMleAttachBetterPartition1` and `kMleAttachBetterPartition2` and implement two separate attempts to a higher priority one?
> A Router MUST make two separate attempts to reconnect to its current Partition, or a higher priority one

File: src/core/thread/mle.cpp [hunk 2]
```diff
@@ -1401,17 +1405,13 @@ ThreadError Mle::SendParentRequest(void)
...
-    SuccessOrExit(error = AppendChallenge(*message, mParentRequest.mChallenge, sizeof(mParentRequest.mChallenge)));
-
     switch (mParentRequestState)
     {
     case kParentRequestRouter:
         scanMask = ScanMaskTlv::kRouterFlag;
 
-        if (mParentRequestMode == kMleAttachSamePartition)
+        if (mParentRequestMode == kMleAttachSamePartition1 ||
+            mParentRequestMode == kMleAttachSamePartition2)
```
Comment (librasungirl) [inline src/core/thread/mle.cpp]: First attempt for `kMleAttachBetterPartition` should also be E/R

Comment (jwhui) [inline src/core/thread/mle.cpp]: Attaching to a better partition only happens in response to receiving an MLE Advertisement that indicates a better partition.  In this case, I think we should first search for Routers and then Routers and REEDs.

Comment (librasungirl) [inline src/core/thread/mle.cpp]: make sense.

**Pull request merged**